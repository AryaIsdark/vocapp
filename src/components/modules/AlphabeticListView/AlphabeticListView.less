@import '~antd/es/style/themes/default.less'; // Import Ant Design styles by less entry

.AlphabeticListView {
  &__container {
    position: relative;
  }

  &__group-header {
    text-transform: uppercase;
    font-size: @headings-font-size-large;
    color: @info-color;
  }

  .ant-list-item {
    display: flex;
    align-items: center;
    padding: 0;
    background: @component-background;

    + .ant-list-item {
      margin-top: @spacing-xx-small;
    }
  }

  .ant-list-item-action {
    flex: 0 0 auto;
    margin-left: 0;
    list-style: none;
    text-align: right;
    padding: @padding-lg;
  }

  .ant-list-item-meta {
    display: flex;
    flex: 1 1;
    align-items: flex-start;
    font-size: 0;
    padding: @padding-lg;
  }

  .ant-list-item-meta-title {
    font-size: @headings-font-size-large;
  }
}
