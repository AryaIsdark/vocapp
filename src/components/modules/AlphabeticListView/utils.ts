const collator = new Intl.Collator('en', {
  numeric: true,
  sensitivity: 'base',
});

export const groupCollectionAlphabetically = <
  T extends Record<string, any>,
  K extends Extract<keyof T, string>
>(
  list: T[],
  key: K,
): Record<string, T[]> => {
  const sortedList = [...list].sort((a: T, b: T) =>
    collator.compare(a[key], b[key]),
  );

  return sortedList.reduce(
    (groupedCollection: Record<string, T[]>, item: T) => {
      const firstCharacter = String(item[key]).charAt(0).toLowerCase();

      if (!groupedCollection[firstCharacter]) {
        groupedCollection[firstCharacter] = [];
      }

      groupedCollection[firstCharacter].push(item);

      return groupedCollection;
    },
    {},
  );
};
