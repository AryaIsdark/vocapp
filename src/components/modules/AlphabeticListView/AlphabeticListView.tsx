import React from "react";
import { Divider } from "antd";
import "./AlphabeticListView.less";
import { groupCollectionAlphabetically } from "./utils";

interface Props<T, K> {
  dataSource: T[];
  indexKey: K;
  onRenderRow: (value: T) => React.ReactElement;
  dataTestId: string;
}

const AlphabeticListView = <T, K extends Extract<keyof T, string>>({
  dataSource,
  indexKey,
  onRenderRow,
  dataTestId,
}: Props<T, K>) => {
  const groupedList = groupCollectionAlphabetically(dataSource, indexKey);
  if (!groupedList) {
    return null;
  }

  return (
    <div className="AlphabeticListView" data-testid={dataTestId}>
      {Object.entries(groupedList).map(([key, value]) => (
        <div key={key} className={"AlphabeticListView__container"}>
          <Divider type={"horizontal"} orientation={"left"}>
            <strong className={"AlphabeticListView__group-header"}>
              {key.toUpperCase()}
            </strong>
          </Divider>
          {value.map(onRenderRow)}
        </div>
      ))}
    </div>
  );
};

export default AlphabeticListView;
