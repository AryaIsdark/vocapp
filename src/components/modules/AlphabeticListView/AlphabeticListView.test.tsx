import React from 'react';
import { render } from '@testing-library/react';
import AlphabeticListView from './AlphabeticListView';

const dataSource = [
  { name: 'arya', account: '12312312' },
  { name: 'booboo', account: '12312312' },
];

const setup = () =>
  render(
    <AlphabeticListView
      onRenderRow={(value: Record<string, any>) => <div key={value.name} />}
      dataSource={dataSource}
      indexKey={'name'}
      dataTestId={'name'}
    />,
  );

describe('components/elements/AlphabeticListView/AlphabeticListView', () => {
  it('renders correctly', () => {
    const { container } = setup();
    expect(container).toMatchSnapshot();
  });
});
