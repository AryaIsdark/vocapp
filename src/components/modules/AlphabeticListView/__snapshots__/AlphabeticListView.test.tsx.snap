// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/elements/AlphabeticListView/AlphabeticListView renders correctly 1`] = `
<div>
  <div
    class="AlphabeticListView"
    data-testid="name"
  >
    <div
      class="AlphabeticListView__container"
    >
      <div
        class="ant-divider ant-divider-horizontal ant-divider-with-text ant-divider-with-text-left"
        role="separator"
      >
        <span
          class="ant-divider-inner-text"
        >
          <strong
            class="AlphabeticListView__group-header"
          >
            a
          </strong>
        </span>
      </div>
      <div />
    </div>
    <div
      class="AlphabeticListView__container"
    >
      <div
        class="ant-divider ant-divider-horizontal ant-divider-with-text ant-divider-with-text-left"
        role="separator"
      >
        <span
          class="ant-divider-inner-text"
        >
          <strong
            class="AlphabeticListView__group-header"
          >
            b
          </strong>
        </span>
      </div>
      <div />
    </div>
  </div>
</div>
`;
